src/gc/gc.cpp

```diff
@@ -899,39 +893,69 @@ class t_join
         }
     }
 
+#ifdef JOIN_STATS
+    uint64_t getTS()
+    {
+        return GCToOSInterface::QueryPerformanceCounter();
+    }
+
+    void startTS (gc_heap* gch)
+    {
+        // parallel execution ends here
+        start[gch->heap_number] = getTS();
+    }
+#endif //JOIN_STATS
+
     void restart()
     {
 #ifdef JOIN_STATS
-        unsigned int elapsed_seq = GetCycleCount32() - start_seq;
-        unsigned int max = 0, sum = 0;
+        uint64_t elapsed_seq = getTS() - start_seq;
+        uint64_t max = 0, sum = 0, wake = 0;
+        uint64_t minTS = start[0];
+        for (int i = 1; i < join_struct.n_threads; i++)
+        {
+            if(minTS > start[i]) minTS = start[i];
+        }
+
         for (int i = 0; i < join_struct.n_threads; i++)
         {
-            unsigned int elapsed = end[i] - start[i];
+            uint64_t wake_delay = start[i] - minTS;
+            uint64_t elapsed = end[i] - start[i];
             if (max < elapsed)
                 max = elapsed;
             sum += elapsed;
+            wake += wake_delay;
         }
-        unsigned int seq_loss = (join_struct.n_threads - 1)*elapsed_seq;
-        unsigned int par_loss = join_struct.n_threads*max - sum;
+        uint64_t seq_loss = (join_struct.n_threads - 1)*elapsed_seq;
+        uint64_t par_loss = join_struct.n_threads*max - sum;
         double efficiency = 0.0;
         if (max > 0)
             efficiency = sum*100.0/(join_struct.n_threads*max);
 
+        const double tsScale = 1e-6;
+
         // enable this printf to get statistics on each individual join as it occurs
-//      printf("join #%3d  seq_loss = %5d   par_loss = %5d  efficiency = %3.0f%%\n", join_id, seq_loss/1000, par_loss/1000, efficiency);
+//      printf("join #%3d  seq_loss = %5g   par_loss = %5g  efficiency = %3.0f%%\n", join_id, tsScale*seq_loss, tsScale*par_loss, efficiency);
 
-        elapsed_total[join_id] += sum/1000;
-        seq_loss_total[join_id] += seq_loss/1000;
-        par_loss_total[join_id] += par_loss/1000;
+        Interlocked::ExchangeAdd(&elapsed_total[id], sum);
+        Interlocked::ExchangeAdd(&wake_total[id], wake);
```

Comment:
no need to change these to Interlocked - restart is only ever called on one thread.